4.14 backport: threads and fork

xavierleroy opened the pull request:
This is a backport to 4.14 of some of the changes in system threads related to the `fork` system call:
- Use a `caml_atfork_hook` instead of `pthread_atfork` to reinitialize the thread machinery after `Unix.fork` (backport of c6d00ee0d44555adb2ec1ce86e8519beff1182c2)
- Remove remaining code related to `pthread_atfork` (backport of 650acfccfc2f09ef88f5acded0e04715e69d6d73)
- Don't destroy locks on I/O channels, this may be dangerous (backport of a change in 5.0 that traces back to Multicore OCaml and that I failed to identify).

Could possibly fix #12636 (to be tested if at all possible).

abbysmal committed:
Introduce an atfork hook in the runtime

This commit introduce an atfork hook in the runtime, as a follow up to the issue encountered in MPR #455 and MPR #471.

As a result, the usage of pthreads_atfork is also removed, opting to rely in both case on manually calling the runtime hook when required.  This way, fork() calls from C code no longer reset the OCaml thread machinery.  This should help with issues such as #12636.

(cherry picked from commit c6d00ee0d44555adb2ec1ce86e8519beff1182c2)
(and adapted to 4.14)

Co-authored-by: xavier.leroy@college-de-france.fr

xavierleroy committed:
Remove st_atfork

The fork hook is now independent from the OS.

Manual cherry-pick from 650acfccfc2f09ef88f5acded0e04715e69d6d73

Co-authored-by: David Allsopp <david.allsopp@metastack.com>

abbysmal reviewed (approved):
Changes are pretty straightforward and just a back port of 5.0 changes: looks good to me.

Considering 5.0's mileage of running problematic programs (like `dune` or `ocaml-lsp`, both exhibiting the same after fork behaviour), I don't think we will see a massive regression here.

We can merge this as is, or maybe wait for a feedback from the reporter at #12636 (though this fix is correct, as destroying potentially locked mutexes is wrong.)

gadmm commented:
The solution for 5.x is therefore to remove the `FIXME` comment about this, or is there another comment/course of action?

xavierleroy commented:
> this fix is correct, as destroying potentially locked mutexes is wrong

But leaving them locked forever is wrong as well :-(   The only standard-conforming implementation that I know of is to lock all the I/O mutexes before forking, then unlock them in both the parent and child processes.  But that can block `Unix.fork` for a long time, waiting for I/O to complete.

A middle ground would be to reinitialize the I/O mutexes in the child process.  This is still nonstandard behavior but perhaps less dangerous than destroying them.  Amusingly, the code started with reinitialization, and switched to destruction in commit https://github.com/ocaml/ocaml/commit/09ba69134f63132a1516dc57c7d17a9c8b1475d9 , and I have no idea why.

abbysmal commented:
Sorry for the late reply.

> A middle ground would be to reinitialize the I/O mutexes in the child process.

This sounds reasonable (glossing over glibc implementation at least, which is not the source of the reported problem anyway.).

Now the question is how consistent are implementation of `pthreads_mutex_init` over cleaning up, and could some trigger assertions if attempting to initialise an already initialised mutex?

TBQH, I think the "not caring about cleaning IO mutexes" behaviour is fine. The runtime already offers support for a highly discouraged practice, in a best effort fashion.

There are other ways to go about it: we could document that people should be careful when `forking` with threads to make sure no thread is holding a lock over a channel, this would ensure the child process can _probably_ destroy them properly.

Or, the right answer, would be to just not tempt the devil by using `fork` in a multithreaded context and write correct code. But this may be naive and I may be missing legitimate use cases for this.

(Sorry for the grumpy tone, I did spend a lot of time working on `dune` running on the early Multicore builds, this brought back not so happy memories :-))

gasche commented:
> We could document that people should be careful when forking with threads to make sure no thread is holding a lock over a channel, this would ensure the child process can probably destroy them properly.

I don't see how users would be able to respect this recommendation if they do use `fork` in a multi-threaded programs. Most I/O operations can schedule a thread out with the mutex locked, the user cannot detect this and I don't see what programming style they could use to reason about whether channels are locked by other threads when they want to `fork()`.

> Or, the right answer, would be to just not tempt the devil by using `fork` in a multithreaded context and write correct code. But this may be naive and I may be missing legitimate use cases for this.

I agree that avoiding a mix of `fork()` and threads is probably the wisest course of action. In older versions of OCaml we would use `fork()` to implement `exec*`-family functions, and this usage has persisted in the ecosystem. (I guess sometimes it still has uses, if you want some cleanup logic to run on the child process before executing.)

abbysmal commented:
> I don't see how users would be able to respect this recommendation if they do use fork in a multi-threaded programs.

Sorry, my statement was a bad attempt at being tongue-in-cheek, it is indeed rather hard (if not impossible) in the most common threading scenarios in OCaml programs. :-)

> (I guess sometimes it still has uses, if you want some cleanup logic to run on the child process before executing.)

Do we have any report of such use cases?

xavierleroy commented:
> >    (I guess sometimes it still has uses, if you want some cleanup logic to run on the child process before executing.)

> Do we have any report of such use cases?

If you look at Jane Street's "spawn" library (https://github.com/janestreet/spawn), you'll see that it does many things in the child process before calling exec.  A good example is to change the working directory for the program that will be exec'd.  It happens that they do all of this (fork, chdir, exec) in C, so once this PR is merged, OCaml's support for fork and threads will no longer interfere with the "spawn" library.  But if someone wanted to write the same kind of custom launcher in OCaml, they would still have to use Unix.fork.

xavierleroy commented:
More generally: in discussing this PR and fork/thread related issues, please distinguish between calling fork() from C and calling Unix.fork () from OCaml.

gasche commented:
Yes, I had this same `chdir` example in mind.

This is a minor point, but I am not fully convinced by the distinction between forking from OCaml and forking from C. I think of `Unix.fork()` as something that will fork *and* ensure that the OCaml runtime system is in a consistent state after forking. If I am calling `fork()` from C, from an OCaml program, and I really want a fork (not exec after a short time), I want to run the same logic to be able to continue executing OCaml code on the child. (It is fine if I need to call a runtime-provided function on the child to do this, instead of having pthread do it magically for me.)

Here it is not just "a `fork()` from C", but more importantly "a `fork()` that will not rely on the OCaml runtime anymore". So I think that the real distinction would be whether we want to "restore" the OCaml runtime on the child, or we are giving up on accessing the runtime. In fact, it would make sense to have an explicit function to call on the child to "give up", instead of hoping that "doing nothing" is always a correct way to do it.

My understanding of the issue at hand is that the 4.x runtime assumes that all `fork()` calls (from C) in an OCaml program want to "restore" the runtime state on the child, and makes implementation choices that are wrong if the intent was to "give up" -- because they error out in cases where restoring is basically impossible but giving up is safe.

xavierleroy committed:
Unix.fork: reinitialize IO channel mutexes in the child instead of destroying them

Some of these mutexes may be locked, so both destroying and
reinitializing are undefined behaviors.  However, destroying can cause
fatal errors in the C pthread library (#12636), while reinitializing
should not.

Fixes: #12636

xavierleroy commented:
I got cold feet about leaving I/O mutexes possibly locked in the child process, as this is a regression w.r.t. 4.14, so I pushed a new version of this PR that reinitializes the mutexes instead of destroying them.

Looking at the sources for Apple's pthread library (https://github.com/apple-oss-distributions/libpthread/tree/libpthread-519), it is clear that pthread_mutex_init doesn't try to check for an already-initialized, already-locked mutex, while pthread_mutex_destroy starts by acquiring the underlying OS lock, and this seems to be causing the fatal error reported in #12636.

> Amusingly, the code started with reinitialization, and switched to destruction in commit https://github.com/ocaml/ocaml/commit/09ba69134f63132a1516dc57c7d17a9c8b1475d9 , and I have no idea why.

I figured it out!  This commit is the one that introduced the st_posix.h and st_win32.h interfaces, instead of calling directly into POSIX or Win32 functions.  These interfaces provide an "allocate and initialize mutex" operation, and a "destroy and free mutex" operation, but no "reinitialize already-allocated mutex" operation.  This PR adds the latter operation; it is a no-op for Win32 since there is no fork() and therefore no use case.

xavierleroy commented:
> My understanding of the issue at hand is that the 4.x runtime assumes that all fork() calls (from C) in an OCaml program want to "restore" the runtime state on the child, and makes implementation choices that are wrong if the intent was to "give up" -- because they error out in cases where restoring is basically impossible but giving up is safe.

That's a good way of looking at it.

xavierleroy commented:
@Engil are you still happy with this PR, now that I/O mutexes are reinitialized after fork?

abbysmal commented:
@xavierleroy Looks good to me, as a bonus it seems to be fixing [the problem witnessed on recent MacOS versions + LSP](https://github.com/ocaml/ocaml/issues/12636#issuecomment-1769070360).

Let's merge it. (need Changes entry fixup)

> These interfaces provide an "allocate and initialize mutex" operation, and a "destroy and free mutex" operation, but no "reinitialize already-allocated mutex" operation. This PR adds the latter operation; it is a no-op for Win32 since there is no fork() and therefore no use case.

It make sense to me, thank you for the digging through the past. 🙂 

> The solution for 5.x is therefore to remove the FIXME comment about this, or is there another comment/course of action?

I think as such the 5.X comment is still pretty much valid and doesn't need to be touched.

xavierleroy committed:
Changes entry for 12646

xavierleroy commented:
Thanks for the feedback.  Now merging!
